### PR TITLE
Collada world exporter now exporting lights

### LINKED ISSUE #901: [Feature] Export lights for collada world exporter
## Desired behavior
Enable capabilities to export lights into the .dae format along with the collada world exporter. 

## Implementation suggestion
Iterate through `ecm.foreach` for each light and 'add' each light to the dae file.

## Additional context
This helps ease the lightmapping process. By maintaining the consistency of lights between ignition and blender, the lighting artist need not tweak lights to sync to the correct positions; and the latest stage of modification will give the final light positions/orientations. This also means ensuring PR #895 exports the light's position/orientations correctly in sdf format. 

I'll be working on this feature on this [branch](https://github.com/ddengster/ign-gazebo/tree/feature/export_lights). Let me know if there's an alternative.

### PR DESCRIPTION
**Pull request by ddengster**
Signed-off-by: ddengster <ed.fan@osrfoundation.org>

# 🎉 New feature

Closes #901 
[Upstream ign-common PR](https://github.com/ignitionrobotics/ign-common/pull/228)

## Summary
Collada world exporter now exporting lights

As seen in ignition
image

As seen in blender
image

## Test it
Run `./build/ignition-gazebo5/bin/INTEGRATION_collada_world_exporter`

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

### CODE REVIEW

File: src/systems/collada_world_exporter/ColladaWorldExporter.cc [hunk 1]
```diff
@@ -36,6 +37,7 @@
 #include <sdf/Visual.hh>
 #include <sdf/Mesh.hh>
 #include <sdf/Model.hh>
+#include <sdf/Light.hh>
```
Comment (ahcorde) [inline src/systems/collada_world_exporter/ColladaWorldExporter.cc:40]: alphabetize

File: src/systems/collada_world_exporter/ColladaWorldExporter.cc [hunk 2]
```diff
@@ -244,9 +246,47 @@ class ignition::gazebo::systems::ColladaWorldExporterPrivate
...
+    [&](const Entity &/*_entity*/,
+        const components::Light *_light,
+        const components::Name *_name)->bool
+    {
+      std::string name = _name->Data();
+      const auto& sdf_light = _light->Data();
+
+      common::ColladaLight p;
+      p.name = name;
+      if (sdf_light.Type() == sdf::LightType::POINT)
```
Comment (ahcorde) [inline src/systems/collada_world_exporter/ColladaWorldExporter.cc:261]: use `{` and `}` for each condition

**Review by ahcorde — CHANGES_REQUESTED** [review]

File: src/systems/collada_world_exporter/ColladaWorldExporter.cc (hunk 2 above)
Comment (ddengster) [reply src/systems/collada_world_exporter/ColladaWorldExporter.cc]: add braces f4fb0a5

File: src/systems/collada_world_exporter/ColladaWorldExporter.cc (hunk 1 above)
Comment (ddengster) [reply src/systems/collada_world_exporter/ColladaWorldExporter.cc]: alphabetize f4fb0a5

**Review by ahcorde — APPROVED** [review]

File: src/systems/collada_world_exporter/ColladaWorldExporter.cc [hunk 3]
```diff
@@ -244,9 +246,48 @@ class ignition::gazebo::systems::ColladaWorldExporterPrivate
...
 
+    std::vector<common::ColladaLight> lights;
+    _ecm.Each<components::Light,
+              components::Name>(
+    [&](const Entity &/*_entity*/,
+        const components::Light *_light,
+        const components::Name *_name)->bool
+    {
+      std::string name = _name->Data();
+      const auto& sdf_light = _light->Data();
```
Comment (chapulina) [inline src/systems/collada_world_exporter/ColladaWorldExporter.cc:257]:
```suggestion
      const auto& sdfLight = _light->Data();
```

File: src/systems/collada_world_exporter/ColladaWorldExporter.cc [hunk 4]
```diff
@@ -244,9 +246,48 @@ class ignition::gazebo::systems::ColladaWorldExporterPrivate
...
+        p.type = "point";
+      } else if (sdf_light.Type() == sdf::LightType::SPOT) {
+        p.type = "spot";
+      } else if (sdf_light.Type() == sdf::LightType::DIRECTIONAL) {
+        p.type = "directional";
+      } else {
+        p.type = "invalid";
+      }
+
+      p.position = sdf_light.RawPose().Pos();
```
Comment (chapulina) [inline src/systems/collada_world_exporter/ColladaWorldExporter.cc:271]: I just noticed that the light struct added to `ign-common` doesn't have the orientation. The convention we follow is that the direction is applied on top of the pose, so we need to use `RawPose().Rot()` somewhere.

See more context in https://github.com/ignitionrobotics/ign-gazebo/issues/329

**Review by chapulina — CHANGES_REQUESTED** [review]

File: src/systems/collada_world_exporter/ColladaWorldExporter.cc (hunk 4 above)
Comment (chapulina) [reply src/systems/collada_world_exporter/ColladaWorldExporter.cc:271]: I think for now we'll have to calculate a new direction using the pose's rotation

File: src/systems/collada_world_exporter/ColladaWorldExporter.cc [hunk 5]
```diff
@@ -244,9 +246,48 @@ class ignition::gazebo::systems::ColladaWorldExporterPrivate
...
+    [&](const Entity &/*_entity*/,
+        const components::Light *_light,
+        const components::Name *_name)->bool
+    {
+      std::string name = _name->Data();
+      const auto& sdf_light = _light->Data();
+
+      common::ColladaLight p;
+      p.name = name;
+      if (sdf_light.Type() == sdf::LightType::POINT) {
```
[the remark this answers is not in the document]
Comment (ddengster) [reply src/systems/collada_world_exporter/ColladaWorldExporter.cc]: style 41e175c

File: src/systems/collada_world_exporter/ColladaWorldExporter.cc (hunk 3 above)
Comment (ddengster) [reply src/systems/collada_world_exporter/ColladaWorldExporter.cc]: style 41e175c

File: src/systems/collada_world_exporter/ColladaWorldExporter.cc (hunk 4 above)
Comment (ddengster) [reply src/systems/collada_world_exporter/ColladaWorldExporter.cc]: apply orientation to local light direction via RotateVec 415d6ac

**Comment by ahcorde** [comment]
@osrf-jenkins retest this please

**Comment by ddengster** [comment]
@chapulina Friendly reminder

**Review by chapulina — APPROVED** [review]
LGTM, thanks for iterating!

**Pull request merged**